torch/distributed/_checkpoint/state_dict_loader.py

```diff
@@ -0,0 +1,234 @@
+    # Note that this is NOT inplace,
+    # it creating a new object and replace what's in the state dict
+    for req in bytes_read_requests:
+        fqn = req.storage_key
+        # Ensure the BytesIO is rewound
+        req.bytes.seek(0)
+        state_dict[fqn] = torch.load(req.bytes)
+
+    tensor_futures.wait()
+
+
+def _validate_sharded_tensor(
+    tensor_md: ShardedTensorMetadata, checkpoint_md: ExtendedTensorMetadata
+) -> List[str]:
+    # We assume the incoming tensor has being validated during construction
+
+    res = []
+    # To ensure a checkpoint can satisfy loading a ST, we compute the loading
+    # plans for all shards and see if they are doable.
+    try:
+        # this API returns a list of issues
+        validate_non_overlapping_shards_metadata(
+            checkpoint_md.tensor_metadata.shards_metadata
+        )
+    except ValueError as e:
+        res.append(str(e))
+
+    for shard_md in tensor_md.shards_metadata:
+        read_volume = 0
+        for storage_md in checkpoint_md.storage_metadata:
+            shard_md_from_storage = storage_md.shard_metadata
+            assert shard_md_from_storage is not None
+            # this is a naive quadratic algo that can later be optimized by
+            #   sorting metadata and the shards md
+            # FIXME what does it mean for offset > 0?
+            assert (
+                storage_md.offset == 0
+            ), f"Cannot handle shard '{shard_md}': offset is non-zero"
+
+            # do they overlap?
+            if not _check_shard_metadata_pair_overlap(shard_md, shard_md_from_storage):
+                continue
+
+            shard_volume = 1
+            for (_, _, _, length,) in _shards_get_overlap_region_wrt_saved_tensor(
+                saved_shard=shard_md_from_storage, current_shard=shard_md
+            ):
+                shard_volume *= length
+            read_volume += shard_volume
+
+        shard_volume = 1
+        for size in shard_md.shard_sizes:
+            shard_volume *= size
+        if read_volume != shard_volume:
+            res.append(
+                f"Shard {shard_md} only has {read_volume} available" +
+                "elements but needs {shard_volume}"
+            )
+    return res
+
+
+def validate_metadata(
+    state_dict: Dict[str, Any], metadata: Metadata
+) -> Optional[List[str]]:
+    """
+    Verify if it's possible to correctly load `state_dict` from `metadata`.
+
+    This method can be used to validate if a checkpoint is usable with a given model.
+
+    Sample Code
+    ```
+        my_model: torch.nn.Model = ....
+        my_reader: torch.distributed._checkpoint.StorageReader = ...
+
+        res = torch.distributed._checkpoint.validate_metadata(my_model.state_dict(), my_reader.read_metadata())
+    ```
+    Args:
+        state_dict: A state_dict to verify if it's loadable.
+        metadata: Checkpoint metadata to verify against.
+
+    Returns:
+        None if no issue was found or a List[str] of issues.
```

Comment:
We should probably just throw an appropriate RuntimeError instead of returning a string of errors like this.